### PR TITLE
Inline replaceAllLiterally and reduce ells

### PR DESCRIPTION
**Pull request by som-snytt**
JDK5 is required, so prefer string.replace.

Go ahead and deprecate the method you can't
say ten times fast.

### CODE REVIEW

File: src/library/scala/collection/StringOps.scala [hunk 1]
```diff
@@ -728,12 +728,13 @@ final class StringOps(private val s: String) extends AnyVal {
...
+    * This method is equivalent to [[java.lang.String#replace(CharSequence,CharSequence)]].
     *
-    *  @param    literal     the string which should be replaced everywhere it occurs
-    *  @param    replacement the replacement string
-    *  @return               the resulting string
+    * @param    literal     the string which should be replaced everywhere it occurs
+    * @param    replacement the replacement string
+    * @return               the resulting string
     */
+  @deprecated("Use `s.replace` as an exact replacement", "2.13.2")
```
Comment (som-snytt) [inline src/library/scala/collection/StringOps.scala:737]: Unintended pun on replacement. But maybe I like it? _Exact replacement_. Get it?

**Comment by martin-g** [comment]
`java.lang.String#replace(java.lang.CharSequence, java.lang.CharSequence)` in JDK 8 is implemented with Regex pattern that is compiled ad-hoc:

```
public String replace(CharSequence target, CharSequence replacement) {
        return Pattern.compile(target.toString(), Pattern.LITERAL).matcher(
                this).replaceAll(Matcher.quoteReplacement(replacement.toString()));
    }
```

I guess this is the reason to have a different implementation. Apache Commons Lang also provides regex-free impl because of the performance gain.

In later versions of Java (not sure which exactly, but 11/12 at least) this method is improved to not use Regex, but it has something else that makes me nervous:

```
public String replace(CharSequence target, CharSequence replacement) {
        String tgtStr = target.toString();
        String replStr = replacement.toString();
        int j = indexOf(tgtStr);
        if (j < 0) {
            return this;
        }
        int tgtLen = tgtStr.length();
        int tgtLen1 = Math.max(tgtLen, 1);
        int thisLen = length();

        int newLenHint = thisLen - tgtLen + replStr.length();
        if (newLenHint < 0) {
            throw new OutOfMemoryError();
        }
        ...
```

It throws **OutOfMemoryError** without a message if the calculated length hint is negative ?!? :bomb:

**Comment by som-snytt** [comment]
@martin-g the scala method was added as a convenience in ancient times https://github.com/scala/scala/blob/2.10.x/src/library/scala/collection/immutable/StringLike.scala#L167

Here, I was going for the shorter method name. I had checked that  the `replaceAll` doesn't allocate if the pattern isn't found, but now I'm surprised to see that pattern compile allocates for literal mode.

I was choosing whether to keep an `indexOf` guard to my usage; I did keep it. There don't seem to be internal usages in anger except `ClassPath` (which could be micro-optimised).

File: src/library/scala/collection/StringOps.scala (hunk 1 above)
Comment (som-snytt) [reply src/library/scala/collection/StringOps.scala:737]: Maybe suggest using a more efficient third-party impl per the comment.

**Comment by som-snytt** [comment]
For the gc-sensitive failing test, https://github.com/scala/scala/pull/8417

**Comment by som-snytt** [comment]
Footnote: if scala wants to improve `string.replace`, maybe it should just hijack string.

**Review by lrytz — APPROVED** [review]

**Review by lrytz — APPROVED** [review]

**Pull request merged**